core/cu29_derive/src/schema.rs

```diff
@@ -0,0 +1,45 @@
+use proc_macro::TokenStream;
+use quote::{quote, ToTokens};
+use syn::{parse_macro_input, DeriveInput};
+
+pub fn derive_schema(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = &input.ident;
+
+    let schema = match &input.data {
+        syn::Data::Struct(data) => {
+            let fields: Vec<_> = data
+                .fields
+                .iter()
+                .map(|f| {
+                    let field_name = f.ident.as_ref().unwrap().to_string();
+                    let field_type = f.ty.to_token_stream().to_string();
+                    quote! {
+                        schema.push((#field_name.to_string(), #field_type.to_string()));
```

Comment:
I think this could cause issues because the same `field_type` can be represented in multiple ways (eg via a type alias or path etc).  Better I think to define a trait:

```rust
pub trait Schema {
    fn schema() -> Vec<(String, String)>;
}
```

and then copper can provide a `derive` macro for ease of implementation, which would delegate to the respective `Schema::schema()` of each field's type (with implementations provided for primitives and library types).

I'm also not entirely convinced about `(String, String)`.  Better I think to have a richer type that can be reliably mapped onto any primitive types known to different schema formats.

Reply:
yes definitely the String->String is not what we want, this is just some quick stuff generated with chatGPT to illustrate the general direction. We should probably go with Value or other enum based stuff?

As for the trait, that is interesting because it will be directly available from the tuple which is pretty cool